Guide/authentication.markdown

```diff
@@ -271,7 +281,78 @@ To create a user with a hashed password, you just need to call the hashing funct
 ```
 
-### Hashing a Password
+The view would look like this, removing the `failedLoginAttempts` that was generated by the IDE, and manually add a password confirmation input field.
+
+```haskell
+-- Web/View/Users/New.hs
+renderForm :: User -> Html
+renderForm user = formFor user [hsx|
+    {(textField #email)}
+    {(passwordField #passwordHash) {fieldLabel = "Password", required = True}}
+
+    <input type="password" name="passwordConfirmation" required/>
+    {submitButton}
+|]
+```
+
+### Editing a User
+
+When editing an existing user we need to special case the password handling. A user may edit their info, but
+without changing their password. In this case we don't want to re-hash the empty input.
+Furthermore, we want to make sure that when we present the form, we don't populate the password field with the hashed password!
+
+```haskell
+-- Web/Controller/Users.hs
+    action UpdateUserAction { userId } = do
+        user <- fetch userId
+        let originalPasswordHash = user.passwordHash
+        -- The value from the password confirmation input field.
+        let passwordConfirmation = param @Text "passwordConfirmation"
+        user
+            |> fill @["email", "passwordHash"]
+            -- We only validate the email field isn't empty, as the password
+            -- can remain empty. We ensure that the error message doesn't include
+            -- the entered password.
+            |> validateField #passwordHash (isEqual passwordConfirmation |> withCustomErrorMessage "Passwords don't match")
+            |> validateField #passwordHash nonEmpty
+            |> validateField #email isEmail
+            -- After this validation, since it's operation on the IO, we'll need to use >>=.
+            |> validateIsUnique #email
+            >>= ifValid \case
+                Left user -> render EditView { .. }
+                Right user -> do
+                    -- If the password hash is empty, then the user did not
+                    -- change the password. So, we set the password hash to
+                    -- the original password hash.
+                    hashed <-
+                        if user.passwordHash == ""
+                            then pure originalPasswordHash
+                            else hashPassword user.passwordHash
+
+                    user <- user
+                        |> set #passwordHash hashed
+                        |> updateRecord
+                    setSuccessMessage "User updated"
+                    redirectTo EditUserAction { .. }
+```
+
+In the case of Editing, the password and password confirmation are optional. If however, the user has changed
+the password, they will also need to confirm it.
+
+
+```haskell
+-- Web/View/Users/Edit.hs
+renderForm :: User -> Html
+renderForm user = formFor user [hsx|
+    {(textField #email)}
+    {(passwordField #passwordHash) {fieldLabel = "Password"}}
+
+    <input type="password" name="passwordConfirmation"/>
```

Comment:
@mpscholten it could have been nicer if we would be able to write

```haskell
{(passwordField #passwordHash) {required = True, fieldLabel = "Password Confirmation", additionalAttributes = [("name", "passwordConfirmation")]}}
```

So the CssFramework would kick in. But now the [additionalAttributs](https://github.com/digitallyinduced/ihp/blob/1bf4fb248f29d9cd65e1b1b7013d34928e202c47/IHP/View/CSSFramework.hs#L112) don't override. 

Do you think they should be able to override?

Reply:
I was able to achieve it with

`{(passwordField #passwordHash) {required = True, fieldLabel = "Password confirmation", fieldName = "passwordConfirmation", validatorResult = Nothing}}`

I'll update the docs

Reply:
While `fieldName` works, I think the `additonalAttributes` should be able to override all the other attributes 👍  can you make a ticket for that?